File: API/src/Modules/Reports/Momentum.Reports.Application/DTOs/ReportDto.cs
using Momentum.Users.Application.DTOs;
using System;

namespace Momentum.Reports.Application.DTOs
{
    public class ReportDto
    {
        public Guid Id { get; set; }
        public Guid SubmitterId { get; set; }
        public Guid? ResolverId { get; set; }
        public UserDto? Submitter { get; set; }
        public UserDto? Resolver { get; set; }
        public DateTime CreatedAt { get; set; }
        public DateTime? UpdatedAt { get; set; }
        public string Data { get; set; }
        public ReportTypeDto Type { get; set; }
        public ReportCategoryDto Category { get; set; }
        public string Message { get; set; }
        public bool Resolved { get; set; }
        public string ResolutionMessage { get; set; }
    }
}

File: libs/backend/dto/src/dtos/report/report.dto.ts
import { Report } from '@momentum/constants';
import { ApiProperty, ApiPropertyOptional, PickType } from '@nestjs/swagger';
import { IsBoolean, IsOptional, IsString } from 'class-validator';
import { UserDto } from '../user/user.dto';
import {
  CreatedAtProperty,
  EnumProperty,
  IdProperty,
  NestedProperty,
  UpdatedAtProperty
} from '../../decorators';
import { ReportCategory, ReportType } from '@momentum/constants';

export class ReportDto implements Report {
  @IdProperty()
  readonly id: number;

  @IdProperty({
    description: 'The ID of the object being referred to by the report',
    bigint: true
  })
  readonly data: number;

  @EnumProperty(ReportType, { description: 'The type of the report' })
  readonly type: ReportType;

  @EnumProperty(ReportCategory, { description: 'The category of the report' })
  readonly category: ReportCategory;

  @ApiProperty({
    description: 'The main text of the report',
    type: String
  })
  @IsString()
  readonly message: string;

  @ApiProperty({
    description: 'Whether the report has been resolved or not',
    type: Boolean
  })
  @IsBoolean()
  readonly resolved: boolean;

  @ApiPropertyOptional({
    description: 'The reason the report was resolved, if it was',
    type: String
  })
  @IsString()
  @IsOptional()
  readonly resolutionMessage: string;

  @IdProperty({ description: 'The user ID of the submitter' })
  readonly submitterID: number;

  @NestedProperty(UserDto)
  readonly submitter: UserDto;

  @IdProperty({
    required: false,
    description: 'The user ID of the resolver, if its been resolved'
  })
  readonly resolverID: number;

  @NestedProperty(UserDto)
  readonly resolver: UserDto;

  @CreatedAtProperty()
  readonly createdAt: Date;

  @UpdatedAtProperty()
  readonly updatedAt: Date;
}

export class CreateReportDto extends PickType(ReportDto, [
  'data',
  'type',
  'category',
  'message'
] as const) {}

export class UpdateReportDto extends PickType(ReportDto, [
  'resolved',
  'resolutionMessage'
] as const) {}
